Add Compression best practices guide (#52968)
dotnet/docs: docs/standard/io/zip-tar-best-practices.md

This article helps you choose the right API, use the convenience methods effectively for trusted input, and safely handle untrusted archives.

## Choose the right API

.NET offers two tiers of archive APIs. Pick the tier that matches your scenario.

### Convenience APIs (one-shot operations)

- `ZipFile.CreateFromDirectory` / `ZipFile.ExtractToDirectory`—create or extract an entire archive in one call.
- `TarFile.CreateFromDirectory` / `TarFile.ExtractToDirectory`—same for TAR.
- Best for: simple workflows with trusted input, quick scripts, build tooling.

### Streaming APIs (entry-by-entry control)

- `ZipArchive`—open an archive, iterate entries, read or write selectively.
- `TarReader` / `TarWriter`—sequential entry-by-entry access.
- Best for: large archives, selective extraction, untrusted input, custom processing.

If you control the archive source (your own build output, known-safe backups), the convenience APIs are the simplest choice. If the archive comes from an external source (user uploads, downloads, network transfers), use the streaming APIs with the safety checks described in this article.

> 

Reviewer: this section is missing ExtractTo* helpers on ZipArchiveEntry or TarEntry

For reference, `TarFile.ExtractToDirectory` validates both the entry path and link target path against the destination directory boundary. If either resolves outside, an `IOException` is thrown. `TarEntry.ExtractToFile()` rejects symbolic and hard link entries entirely—it throws `InvalidOperationException`.

### Complete safe extraction examples

Combine path traversal validation, size limits, entry count limits, and link handling in a single extraction loop.

#### ZIP—complete safe extraction

The following method extracts a ZIP archive while enforcing all recommended safety checks:

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractZip":::

#### TAR—complete safe extraction

TAR extraction differs from ZIP in several ways: entries are read sequentially (there's no central directory), link entries need explicit handling, and the `DataStream` must be consumed before advancing to the next entry.

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractTar":::

## Memory and performance considerations

### ZipArchiveMode.Update loads entries into memory

> Don't use `ZipArchiveMode.Update` for large or untrusted archives. When you open a `ZipArchive` in `Update` mode and call `Open()` or `OpenAsync()` on an entry, its uncompressed data is loaded into a `MemoryStream` to support in-place modifications. Accessing entry metadata (such as `FullName`, `Length`, or `ExternalAttributes`) does not trigger decompression. For large or malicious archives, opening entry content streams can cause `OutOfMemoryException`.

Reviewer: It would help to mention here that users should check `Length` before opening the entry

> Additionally, when you open a `ZipArchive` in `Read` mode with an **unseekable** stream (for example, a network stream), the runtime copies the entire stream into a `MemoryStream` up front to enable seeking through the central directory.

Reviewer: ```suggestion
Additionally, when you open a `ZipArchive` in `Read` mode with an **unseekable** stream (for example, a network stream), the runtime buffers the entire archive contents in memory to enable seeking through the central directory.
```

Reviewer: Also, this sentence is in `ZipArchiveMode.Update loads entries into memory` subsection, but talks about `Read` mode, so some restructuring is needed (maybe remove the subsection heading?)

## Platform considerations

### Unix file permissions

- **ZIP:** Unix permissions are stored in the upper 16 bits of `ExternalAttributes`. When extracting on Unix via `ExtractToDirectory` or `ExtractToFile`, the runtime restores ownership permissions (read/write/execute for user/group/other), subject to the process umask. SetUID, SetGID, and StickyBit are stripped. Permissions are not applied if the upper bits are zero. This happens when the ZIP was created on Windows, because the Windows runtime sets `DefaultFileExternalAttributes` to `0`. On Windows, these attributes are always ignored during extraction.
- **TAR:** The `TarEntry.Mode` property represents `UnixFileMode` and can store all 12 permission bits (read/write/execute for user/group/other, plus SetUID, SetGID, and StickyBit). When extracting on Unix via `ExtractToDirectory` or `ExtractToFile`, the runtime applies only the 9 ownership bits (rwx for user/group/other), subject to the process umask. SetUID, SetGID, and StickyBit are stripped for security.

> When processing untrusted archives, validate `TarEntry.Mode` before extracting. An archive could set executable permissions on files that should not be executable.

Reviewer: The examples above don't touch/apply `Mode` at all, that is probably related to the lack of mention of `ExtractToFile` helpers in the earlier parts of the article.

Reviewer: > An archive could set executable permissions on files that should not be executable.

That sentence is weird, the archive author clearly may have wanted some files to be executable, so who really decides that the files `should not` be executable?

I would reword it simply as a warning along the lines that untrusted archives may contain malicious executable files

### Special entry types (TAR)

Block devices, character devices, and FIFOs can only be created on Unix. Extracting these on Windows throws an exception. Elevated privileges are required to create block and character device entries.

### File name sanitization differs by platform

> On Windows, when using `ExtractToDirectory`, the runtime replaces control characters and ``"*:<>?|`` with underscores in entry names. On Unix, only null characters are replaced. Archive entries with names like `file:name.txt` are renamed to `file_name.txt` on Windows but extracted as-is on Unix. The per-entry APIs (`Open()`, `ExtractToFile()`) do not perform any name sanitization.

Reviewer: > The per-entry APIs (`Open()`, `ExtractToFile()`) do not perform any name sanitization.

Do you think that we should mention `entry.ExtractToFile(entry.Name)` is safe only if user properly validates Name/Mode/...?

- **TAR** (`System.Formats.Tar`): A Unix-origin archive format that stores files, directories, and metadata (permissions, ownership, timestamps) without compression. .NET supports the V7, UStar, PAX, and GNU formats. The primary types are <xref:System.Formats.Tar.TarReader> and <xref:System.Formats.Tar.TarWriter> for streaming access, and <xref:System.Formats.Tar.TarFile> for file-based convenience methods. TAR is often combined with a compression layer (for example, `GZipStream` for `.tar.gz` files).

This article helps you choose the right API, use the convenience methods effectively for trusted input, and safely handle untrusted archives.

## Choose the right API

.NET offers two tiers of archive APIs. Pick the tier that matches your scenario.

### Convenience APIs (one-shot operations)

> - `ZipFile.CreateFromDirectory` / `ZipFile.ExtractToDirectory`—create or extract an entire archive in one call.

Reviewer 2: I think most of the `` `code` `` in this article should be API reference links instead.

> [!CAUTION]
> A ZIP archive primarily transmits files, while a TAR archive transmits a filesystem topology, including file types, symbolic links, hard links, permissions, and other metadata. This gives the TAR extraction process much more control over how data is represented on disk. Because these structures are meaningful to the filesystem, an adversary can influence security-impacting behaviors beyond filenames and file contents. Exercise extra caution when processing untrusted TAR archives.

## Work with trusted archives

When the archive source is known and trusted, the convenience methods give you a safe, one-line extraction path:

- `ZipFile.ExtractToDirectory` and `TarFile.ExtractToDirectory` handle path validation automatically. They sanitize entry names, resolve paths, and check directory boundaries.
- `ZipFile.ExtractToDirectory` has overloads that default to not overwriting existing files. All `TarFile.ExtractToDirectory` overloads require the `overwriteFiles` parameter, so you must always choose explicitly.

> - When overwriting is enabled during ZIP extraction, .NET extracts to a temporary file first and only replaces the target after successful extraction. This prevents partial corruption if the extraction fails.

Reviewer 2: What happens when you allow overwriting existing files with TAR? If it's less of a concern with TAR, because it's not compressed, is that worth mentioning here?

Author: Good catch, I will write a comment that a backup copy is advised, since the initial file gets lost

> [!NOTE]
> The convenience methods don't enforce size limits, entry count limits, or other policies needed for safe extraction of untrusted archives. If that matters even for trusted input (for example, very large archives), use the streaming approach described in [Handle untrusted archives safely](#handle-untrusted-archives-safely).

## Handle untrusted archives safely

For untrusted input—user uploads, third-party downloads, or network transfers—iterate over entries manually and enforce your own safety checks. The following subsections describe what you need to enforce and why.

### What the convenience methods don't protect you from

> `ExtractToDirectory` handles path traversal validation (including symbolic link targets in TAR), but it doesn't enforce size limits or entry count limits. A small compressed file can expand to terabytes of data (known as a *zip bomb*). You must enforce these limits yourself when processing untrusted input.

Reviewer 2: To me, it's not clear what "path traversal validation" means exactly. Is it worth expanding on that? And similar for "directory boundary" later in the article.

### Enforce size and entry count limits

Neither `ZipArchive` nor `TarReader` limits the total uncompressed size or the number of entries extracted, and neither do the `ExtractToDirectory` convenience methods. You must enforce these limits yourself.

> [!IMPORTANT]

> > A small compressed file can expand to terabytes of data—this is known as a *zip bomb*. Always enforce limits on decompressed size and entry count when extracting untrusted archives.

Reviewer 2: This just repeats what was said few paragraphs above. Should one of the two duplicate sentences be removed?

#### Enforce per-entry size limits

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractEntry":::

#### Track aggregate size and entry count

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractArchive":::

> [!TIP]
> The same approach applies to TAR archives. Since TAR files are read entry-by-entry via `TarReader.GetNextEntry()`, track both the cumulative data size and entry count as you iterate.

### Validate destination paths

> When you use the streaming APIs, you're responsible for validating every entry's destination path. The low-level APIs perform no path validation at all.

Reviewer 2: This is the only mention of "low-level APIs" in this article. I don't think it makes sense to use this term here.

- The trailing directory separator on `fullDestDir` is critical—without it, a path like `/safe-dir-evil/file` would incorrectly match `/safe-dir`.

> [!WARNING]
> The following APIs leave you completely unprotected against path traversal. You must validate paths yourself before calling them.

- `ZipArchiveEntry.ExtractToFile()` writes to whatever path you give it—no sanitization, no boundary check.
- `ZipArchiveEntry.Open()` returns a raw `Stream`—the caller decides where to write.
- `TarEntry.ExtractToFile()` writes to the given path without validating it against any directory boundary.

**Vulnerable pattern—DO NOT USE without validation:**

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="VulnerablePattern":::

### Handle symbolic and hard links (TAR)

TAR archives support symbolic links and hard links, which introduce attack vectors beyond basic path traversal:

- **Symlink escape:** A symlink entry points to an arbitrary location (for example, `/etc/`), then a subsequent file entry relative to the symlink directory writes through the link to that external location.

> - **Hard link to sensitive file:** A hard link target references a file outside the extraction directory. Because a hard link shares the same inode as the original, any code that later opens the hard link for writing modifies the original file's contents. Simply overwriting the hard link (for example, with `File.Create`) replaces the directory entry and does not affect the original.

Reviewer 2: > Simply overwriting the hard link (for example, with `File.Create`) replaces the directory entry and does not affect the original.

I don't understand why this sentence is here. Also, it doesn't seem to be true: calling `File.Create` on an existing hardlink will truncate and write to the original file.

Author: You're right, I iwll update this

The safest approach for untrusted archives is to skip link entries entirely:

```csharp
if (entry.EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
    continue; // Skip link entries for untrusted input
```

If you need to preserve links, validate that the link target resolves within your destination directory before creating it:

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="ValidateSymlink":::

> If your use case requires hard links but you want to avoid filesystem-level hard links, `TarHardLinkMode.CopyContents` copies the file content instead of creating a real hard link. This eliminates hard-link-based attacks and produces more portable output on Windows.

Reviewer 2: This could be understood as somehow creating hardlinks that are not filesystem-level.  I would rephrase it to something like:

```suggestion
If your use case requires extracting archives with hard links but you want to avoid hard links on disk, `TarHardLinkMode.CopyContents` copies the file content instead of creating a hard link. This eliminates hard-link-based attacks and produces more portable output on Windows.
```

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractTar":::

## Memory and performance considerations

### ZipArchive memory usage

Don't use `ZipArchiveMode.Update` for large or untrusted archives. When you open a `ZipArchive` in `Update` mode and call `Open()` or `OpenAsync()` on an entry, its uncompressed data is loaded into a `MemoryStream` to support in-place modifications. Accessing entry metadata (such as `FullName`, `Length`, or `ExternalAttributes`) does not trigger decompression. For large or malicious archives, opening entry content streams can cause `OutOfMemoryException`. Check `ZipArchiveEntry.Length` before calling `Open()` to avoid decompressing unexpectedly large entries.

Additionally, when you open a `ZipArchive` in `Read` mode with an **unseekable** stream (for example, a network stream), the runtime buffers the entire archive contents in memory to enable seeking through the central directory.

```csharp
// Update mode: calling entry.Open() loads the full entry into memory
using var archive = new ZipArchive(stream, ZipArchiveMode.Update);

> ```

Reviewer 2: I think this snippet is unnecessary.

**Recommendation:** Only use `Update` mode for archives you trust and know are small enough to fit in memory. For large archives, create a new archive and selectively copy entries:

> :::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="StreamingApproach":::

Reviewer 2: > For large archives, create a new archive and selectively copy entries

I think it's not completely clear whether "selectively copy entries" is here to achieve the same result as if we used `Update` or to avoid copying large entries. I would add a note here explicitly saying that `Read` and `Create` are safer.

### TAR streaming model

> `TarReader` reads entries one at a time and doesn't buffer the entire archive. However, for unseekable streams, each entry's `DataStream` is only valid until the next `GetNextEntry()` call. If you need to retain entry data, either copy it immediately or pass `copyContents: true` to `GetNextEntry()`, which copies the entry data into a separate `MemoryStream` that remains valid after advancing:

Reviewer 2: `copyContents: true` has similar memory considerations as `ZipArchiveMode.Update`, right? Is that worth mentioning here?

Author: Yes, I think so, I'll mention it too

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="TarStreaming":::

### Thread safety

> `ZipArchive`, `TarReader`, and `TarWriter` are not thread-safe. Don't access an instance from multiple threads concurrently. If you need parallel processing, use a separate instance per thread or synchronize access externally.

Reviewer 2: This could be understood as: "if you want to write to an archive from multiple threads, have a separate `ZipArchive`/`TarWriter` on each thread". But that wouldn't work correctly, no?

Author: Makes sense, I will reprhase it

## Platform considerations

### Unix file permissions

> - **ZIP:** Unix permissions are stored in the upper 16 bits of `ExternalAttributes`. When extracting on Unix via `ExtractToDirectory` or `ExtractToFile`, the runtime restores ownership permissions (read/write/execute for user/group/other), subject to the process umask. SetUID, SetGID, and StickyBit are stripped. Permissions are not applied if the upper bits are zero. This happens when the ZIP was created on Windows, because the Windows runtime sets `DefaultFileExternalAttributes` to `0`. On Windows, these attributes are always ignored during extraction.

Reviewer 2: I would not use the phrase "Windows runtime" here, especially since Windows Runtime is something completely different.

```suggestion
- **ZIP:** Unix permissions are stored in the upper 16 bits of `ExternalAttributes`. When extracting on Unix via `ExtractToDirectory` or `ExtractToFile`, the runtime restores ownership permissions (read/write/execute for user/group/other), subject to the process umask. SetUID, SetGID, and StickyBit are stripped. Permissions are not applied if the upper bits are zero. This happens when the ZIP was created on Windows, because .NET on Windows sets `DefaultFileExternalAttributes` to `0`. On Windows, these attributes are always ignored during extraction.
```

> [!NOTE]
> In versions earlier than .NET 11, the runtime didn't validate ZIP CRC-32 values on read. The runtime computed CRC-32 values when writing ZIP entries for storage in the archive, but didn't verify them during extraction. If you target a runtime earlier than .NET 11, corrupt or tampered ZIP entries might be accepted silently.

> [!NOTE]
> CRC-32 isn't a cryptographic hash—it detects accidental corruption but doesn't protect against intentional tampering by a sophisticated attacker.

## Untrusted metadata

### ZIP comments and extra fields

- **Archive and entry comments** are arbitrary strings. If your application displays or processes comments, sanitize them appropriately.
- **Extra fields** are binary key-value pairs attached to each entry. The runtime preserves unknown extra fields and trailing data when reading and writing archives in `Update` mode and round-trips them as-is. If your application reads or interprets extra fields, validate their contents.

> - **Entry name encoding:** when writing, the runtime uses ASCII for entry names that contain only printable characters (32-126) and UTF-8 (with the language encoding flag set) for names that contain other characters. When reading without a custom encoding, entries with the language encoding flag are decoded as UTF-8, and entries without it are also decoded as UTF-8. Use the `entryNameEncoding` parameter on `ZipArchive` to override encoding when needed, but be aware the override affects all entries uniformly.

Reviewer 2: Nit about ASCII printable characters, I found the phrasing a bit odd and not everyone might know that UTF-8 is compatible with ASCII:

```suggestion
- **Entry name encoding:** when writing, the runtime uses ASCII for entry names that contain only ASCII printable characters (32-126) and UTF-8 (with the language encoding flag set) for names that contain other characters. When reading without a custom encoding, entries with or without the language encoding flag are decoded as UTF-8 (which also correctly decodes ASCII). Use the `entryNameEncoding` parameter on `ZipArchive` to override encoding when needed, but be aware the override affects all entries uniformly.
```

## Encryption considerations (.NET 11+)

> [!NOTE]
> ZIP encryption support (ZipCrypto and WinZip AES) is new in .NET 11.

.NET 11 adds support for reading and writing encrypted ZIP archives using WinZip-compatible encryption. The `ZipEncryptionMethod` enum specifies the encryption method:

| Value | Description |
|-------|-------------|
| `None` | No encryption. |
| `ZipCrypto` | Legacy ZIP encryption. Use only for backward compatibility—vulnerable to known-plaintext attacks. |
| `Aes128` | WinZip AES-128. |
| `Aes192` | WinZip AES-192. |
| `Aes256` | WinZip AES-256. **Recommended**—strongest available option. |
| `Unknown` | Returned when the entry uses an encryption method that .NET does not support. |

### Choose AES-256 for new archives

When creating encrypted entries, always prefer `Aes256`. `ZipCrypto` is a legacy method with known cryptographic weaknesses and shouldn't be relied upon for security—use it only when interoperating with tools that don't support WinZip AES.

```csharp
// ⚠️ Weak encryption — use only for backward compatibility
archive.CreateEntry("file.txt", "password", ZipEncryptionMethod.ZipCrypto);

// ✅ Prefer AES-256

> archive.CreateEntry("file.txt", "password", ZipEncryptionMethod.Aes256);

Reviewer 2: It's quite obvious, but in case somebody copies this code and doesn't change the password, I would turn password into a variable:

```suggestion
// ⚠️ Weak encryption — use only for backward compatibility
archive.CreateEntry("file.txt", password, ZipEncryptionMethod.ZipCrypto);

// ✅ Prefer AES-256
archive.CreateEntry("file.txt", password, ZipEncryptionMethod.Aes256);
```

```

### Read encrypted entries

Use `ZipArchiveEntry.EncryptionMethod` to check the encryption method, and provide a password to `Open`:

```csharp
using ZipArchive archive = ZipFile.OpenRead("encrypted.zip");

foreach (ZipArchiveEntry entry in archive.Entries)
{
    if (entry.EncryptionMethod == ZipEncryptionMethod.Unknown)
    {
        // Unsupported encryption method, skip this entry
        continue;
    }

    using Stream stream = entry.Open("myPassword");
    // ... read the decrypted data
}
```

Attempting to open an entry that uses an unsupported encryption method (`ZipEncryptionMethod.Unknown`) throws `NotSupportedException`.

### Convenience methods with encryption

New option types let you pass a password and encryption method to the convenience APIs:

```csharp
// Extract an encrypted archive
ZipFile.ExtractToDirectory("encrypted.zip", destDir, new ZipExtractionOptions
{

>     Password = "myPassword".AsMemory(),
>     OverwriteFiles = false
> });
> 
> // Create an encrypted archive
> ZipFile.CreateFromDirectory(sourceDir, "encrypted.zip", new ZipFileCreationOptions
> {
>     Password = "myPassword".AsMemory(),

Reviewer 3: ```suggestion
    Password = "myPassword",
    OverwriteFiles = false
});

// Create an encrypted archive
ZipFile.CreateFromDirectory(sourceDir, "encrypted.zip", new ZipFileCreationOptions
{
    Password = "myPassword",
```

Author: The argument is right, we are passsing readonlymemory<char>.

    EncryptionMethod = ZipEncryptionMethod.Aes256,
    CompressionLevel = CompressionLevel.Optimal
});
```

## Security checklist

Before deploying code that handles archives from untrusted sources, verify you've addressed each of the following:

> - [ ] **Manual iteration:** Don't use `ExtractToDirectory` for untrusted input—iterate entries manually to enforce all limits.

Reviewer 2: The GitHub checklist syntax doesn't seem to work here, so just use regular bullet points.

- [ ] **Path traversal:** Validate all destination paths with `Path.GetFullPath()` + `StartsWith()`.
- [ ] **Decompression bombs:** Enforce limits on decompressed size (per-entry and total) and entry count.
- [ ] **Symlink/hardlink attacks (TAR):** Validate link targets resolve within the destination, or skip link entries entirely.
- [ ] **Memory limits:** Avoid `ZipArchiveMode.Update` for large untrusted archives. Avoid `Read` mode with unseekable streams from untrusted sources.
- [ ] **Thread safety:** Don't share `ZipArchive`, `TarReader`, or `TarWriter` instances across threads.
- [ ] **Untrusted metadata:** Treat entry names, comments, and extra fields as untrusted input. Sanitize before display or processing.
- [ ] **Overwrite behavior:** Default to `overwrite: false`.

> - [ ] **Resource disposal:** Always dispose `ZipArchive`, `TarReader`, `TarWriter`, and their streams.

Reviewer 3: Nit: Thread safety and Resource disposal feel a bit out of place in a "Security checklist".

### Convenience APIs (one-shot operations)

- `ZipFile.CreateFromDirectory` / `ZipFile.ExtractToDirectory`—create or extract an entire archive in one call.
- `TarFile.CreateFromDirectory` / `TarFile.ExtractToDirectory`—same for TAR.
- Best for: simple workflows with trusted input, quick scripts, build tooling.

### Streaming APIs (entry-by-entry control)

- `ZipArchive`—open an archive, iterate entries, read or write selectively.
- `TarReader` / `TarWriter`—sequential entry-by-entry access.
- Best for: large archives, selective extraction, untrusted input, custom processing.

If you control the archive source (your own build output, known-safe backups), the convenience APIs are the simplest choice. If the archive comes from an external source (user uploads, downloads, network transfers), use the streaming APIs with the safety checks described in this article.

## Work with trusted archives

When the archive source is known and trusted, the convenience methods give you a safe, one-line extraction path:

- `ZipFile.ExtractToDirectory` and `TarFile.ExtractToDirectory` handle path validation automatically—they sanitize entry names, resolve paths, and check directory boundaries.

> - Default overwrite behavior is `false`. Always be explicit:

Reviewer: If the default is safe, why would being explicit about it be better?

Also, only ZipFile.ExtractToDirectory has overload without `overWriteFiles`

```csharp
// ✅ Explicit — default is safe
ZipFile.ExtractToDirectory("archive.zip", destDir, overwriteFiles: false);
TarFile.ExtractToDirectory("archive.tar", destDir, overwriteFiles: false);
```

- When overwriting is enabled during ZIP extraction, .NET extracts to a temporary file first and only replaces the target after successful extraction—this approach prevents partial corruption if the extraction fails.

> [!NOTE]
> The convenience methods don't limit decompressed size or entry count. If that matters even for trusted input (for example, very large archives), use the streaming approach described in [Handle untrusted archives safely](#handle-untrusted-archives-safely).

## Handle untrusted archives safely

> For untrusted input—user uploads, third-party downloads, or network transfers—iterate over entries manually and enforce your own safety checks. The following subsections describe what you need to enforce and why.

Reviewer: ```suggestion
For untrusted input—user uploads, third-party downloads, or network transfers: iterate over entries manually and enforce your own safety checks. The following subsections describe what you need to enforce and why.
```

### What the convenience methods don't protect you from

> `ExtractToDirectory` handles path traversal validation, but it doesn't enforce size limits, entry count limits, or filter dangerous TAR entry types. A small compressed file can expand to terabytes of data (known as a *zip bomb*), and TAR archives can contain symbolic links that escape the extraction directory. You must handle these yourself when processing untrusted input.

Reviewer: > TAR archives can contain symbolic links that escape the extraction directory. You must handle these yourself when processing untrusted input.

but we *do* check link destination in ExtractToDirectory.

### Enforce size and entry count limits

Neither `ZipArchive` nor `TarReader` limits the total uncompressed size or the number of entries extracted, and neither do the `ExtractToDirectory` convenience methods. You must enforce these limits yourself.

> [!IMPORTANT]
> A small compressed file can expand to terabytes of data—this is known as a *zip bomb*. Always enforce limits on decompressed size and entry count when extracting untrusted archives.

#### Track decompressed size during extraction

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractEntry":::

#### Track aggregate size and entry count

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractArchive":::

> [!TIP]
> The same approach applies to TAR archives. Since TAR files are read entry-by-entry via `TarReader.GetNextEntry()`, track both the cumulative data size and entry count as you iterate.

> ### Validate destination paths (low-level APIs only)

Reviewer: ```suggestion
### Validate destination paths
```

All headers below relate to per-entry APIs

When you use the streaming APIs, you're responsible for validating every entry's destination path. The low-level APIs perform no path validation at all.

For every entry, resolve the destination to an absolute path and verify it falls within your target directory:

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="PathValidation":::

Key points:

- `Path.GetFullPath()` resolves relative segments like `../` into an absolute path.
- The `StartsWith` check ensures the resolved path is still inside the destination.
- The trailing directory separator on `fullDestDir` is critical—without it, a path like `/safe-dir-evil/file` would incorrectly match `/safe-dir`.

> [!NOTE]

> > `ExtractToDirectory` handles path traversal for you—the runtime sanitizes entry names, resolves paths with `Path.GetFullPath()`, and verifies them with `StartsWith`. You're using the streaming APIs here because of the size-limits issue described above.

Reviewer: I am not sure what this note is supposed to achieve here, we are not talking about `ExtractToDirectory` here

> [!WARNING]
> The following APIs leave you completely unprotected against path traversal. You must validate paths yourself before calling them.

- `ZipArchiveEntry.ExtractToFile()` writes to whatever path you give it—no sanitization, no boundary check.
- `ZipArchiveEntry.Open()` returns a raw `Stream`—the caller decides where to write.
- `TarEntry.ExtractToFile()` writes to the given path without validating it against any directory boundary.

**Vulnerable pattern—DO NOT USE without validation:**

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="VulnerablePattern":::

### Handle symbolic and hard links (TAR)

TAR archives support symbolic links and hard links, which introduce attack vectors beyond basic path traversal:

- **Symlink escape:** A symlink entry points to an arbitrary location (for example, `/etc/`), then a subsequent file entry relative to the symlink directory writes through the link to that external location.

> - **Hard link to sensitive file:** A hard link target references a file outside the extraction directory, allowing reads or overwrites.

Reviewer: creating a hardlink to a file and then overwriting the hardlink with a different file does not affect the original file.

That happens only if you open the file in RW mode and rewrite the contents.

For reference, `TarFile.ExtractToDirectory` validates both the entry path and link target path against the destination directory boundary. If either resolves outside, an `IOException` is thrown. `TarEntry.ExtractToFile()` rejects symbolic and hard link entries entirely—it throws `InvalidOperationException`.

### Complete safe extraction examples

Combine path traversal validation, size limits, entry count limits, and link handling in a single extraction loop.

#### ZIP—complete safe extraction

The following method extracts a ZIP archive while enforcing all recommended safety checks:

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractZip":::

#### TAR—complete safe extraction

TAR extraction differs from ZIP in several ways: entries are read sequentially (there's no central directory), link entries need explicit handling, and the `DataStream` must be consumed before advancing to the next entry.

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="SafeExtractTar":::

## Memory and performance considerations

### ZipArchiveMode.Update loads entries into memory

> Don't use `ZipArchiveMode.Update` for large or untrusted archives. When you open a `ZipArchive` in `Update` mode, each entry's uncompressed data is loaded into a `MemoryStream` when that entry is accessed. The runtime requires a seekable stream for Update mode and decompresses each entry fully into memory to support in-place modifications. For large or malicious archives, this behavior can cause `OutOfMemoryException`.

Reviewer: The content is decompressed when user opens the content stream via ZipArchiveEntry.Open(Async) only, accessing the metadata is safe.

Additionally, when you open a `ZipArchive` in `Read` mode with an **unseekable** stream (for example, a network stream), the runtime copies the entire stream into a `MemoryStream` up front to enable seeking through the central directory.

```csharp

> // ⚠️ In Update mode, each entry is decompressed into memory when accessed

Reviewer: the exclamation mark at the start is confusing. It seems to warn that the following code is unsafe, but when user reads to the end of the line there is the "when accessed" (which may not be clear what it means, see comment above), so the code is safe?

It still leaves open questions for the reader IMO

using var archive = new ZipArchive(stream, ZipArchiveMode.Update);
```

**Recommendation:** Only use `Update` mode for archives you trust and know are small enough to fit in memory. For large archives, create a new archive and selectively copy entries:

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="StreamingApproach":::

### TAR streaming model

> `TarReader` reads entries one at a time and doesn't buffer the entire archive. However, for unseekable streams, each entry's `DataStream` is only valid until the next `GetNextEntry()` call. If you need to retain entry data, copy it immediately:

Reviewer: there is optional `copyContents` parameter in `GetNextEntry which lets users retain entry data.

:::code language="csharp" source="./snippets/zip-tar-best-practices/csharp/Program.cs" id="TarStreaming":::

### Thread safety

`ZipArchive` isn't thread-safe. The internal state—entry lists, stream positions, and disposal flags—isn't synchronized. Don't read or write entries from multiple threads concurrently. If you need parallel processing, open a separate `ZipArchive` instance per thread, or synchronize access externally.

> `TarReader` and `TarWriter` are likewise not designed for concurrent use. Each operates on a single underlying stream with sequential access semantics.

Reviewer: This may be summarized more succintly by something along the lines "None of the classes are threadsafe"., we don't need to mention the technical details why they are not threadsafe

- **TAR:** The `TarEntry.Mode` property represents `UnixFileMode` and can store all 12 permission bits (read/write/execute for user/group/other, plus SetUID, SetGID, and StickyBit). However, during **regular file extraction**, only the 9 ownership bits (rwx for user/group/other) are applied—SetUID, SetGID, and StickyBit are explicitly stripped for security. Directories, block devices, character devices, and FIFOs receive the full `Mode` value including SetUID, SetGID, and StickyBit.

### Special entry types (TAR)

Block devices, character devices, and FIFOs can only be created on Unix. Extracting these on Windows throws an exception. Elevated privileges are required to create block and character device entries.

### File name sanitization differs by platform

> On Windows, the runtime replaces control characters and `"*:<>?|` with underscores via `ArchivingUtils.SanitizeEntryFilePath()`. On Unix, only null characters are replaced. Archive entries with names like `file:name.txt` are renamed to `file_name.txt` on Windows but extracted as-is on Unix.

Reviewer: mention to `ArchivingUtils` is an implementation detail. What APIs does this apply to? only to ExtractToDirectory?

## Data integrity

> ZIP entries include a CRC-32 checksum that you can use to verify data hasn't been corrupted or tampered with.

Reviewer: We also started checking CRC32 in TAR headers

Reviewer: I think you should also mention TAR CRC in the first paragraph, now users might assume that this section is Zip-only and skip the rest of it.

> [!NOTE]
> In prior versions of .NET, no CRC validation was performed on read. The runtime computed CRC values when writing entries (for storage in the archive), but never verified them during extraction. If you're targeting a runtime older than .NET 11, be aware that corrupt or tampered entries are silently accepted.

> [!NOTE]
> CRC-32 isn't a cryptographic hash—it detects accidental corruption but doesn't protect against intentional tampering by a sophisticated attacker.

## Untrusted metadata

### ZIP comments and extra fields

Treat all archive metadata as untrusted input. ZIP archives can contain attacker-controlled metadata beyond entry names:

- **Archive and entry comments** are arbitrary strings encoded using either Code Page 437 or UTF-8 (depending on the language encoding flag). If your application displays or processes comments, sanitize them appropriately.
- **Extra fields** are binary key-value pairs attached to each entry. The runtime preserves unknown extra fields and trailing data when reading and writing archives in `Update` mode—they're round-tripped as-is. If your application reads or interprets extra fields, validate their contents.

> - **Entry name encoding** defaults to the system codepage for entries without the language encoding flag (EFS) set, and UTF-8 when EFS is set. The ZIP specification defines Code Page 437 as the default, but in practice, most tools (including the Windows Shell zip tool) use the local system codepage instead, and .NET follows the same behavior. When interoperating with archives from other tools, mismatched encodings can produce garbled file names. Use the `entryNameEncoding` parameter on `ZipArchive` to override encoding when needed, but be aware the override affects all entries uniformly.

Reviewer: > .NET Follows the same behavior

I thought we default to ASCII, and not current code-page

Author: Yes, that's right, I'll change it

### TAR header-driven memory allocation

TAR entry headers contain size fields that the parser uses to allocate buffers. A malicious TAR archive can declare an extremely large size for a PAX extended attributes block or a GNU long-path entry, causing the parser to attempt a large memory allocation. The runtime does include a `ValidateSize()` guard that rejects allocations exceeding `Array.MaxLength` (~2 GB), so allocations aren't completely unbounded—but values up to ~2 GB can still cause significant memory pressure. Your entry-count and per-entry-size limits (described in [Enforce size and entry count limits](#enforce-size-and-entry-count-limits)) also help mitigate this risk, since these metadata entries are counted and sized like regular entries.

> ## Encryption considerations (preview)

Reviewer: ```suggestion
## Encryption considerations (.NET 11+)
```

We are committing to ship encryption in .NET 11, so if you want to keep this section, either remove the (preview) or make it clear that it only applies to future .NET version.

> [!NOTE]

> > ZIP encryption support (ZipCrypto and WinZip AES) is a preview feature that isn't yet publicly available. The APIs described in this section are subject to change.

Reviewer: `preview` feature generally means something that we shipped, but requires explicit opt-in because we do not consider the API to be stable and reserve the right to change it between releases. E.g. we shipped QUIC APIs as preview in .NET 7 (IIRC) and removed the preview label in subsequent releases).

Since we did not ship encryption support yet, it feels weird to me to call it a `preview` feature.

.NET 11 adds support for reading and writing encrypted ZIP archives using WinZip-compatible encryption. The `ZipEncryptionMethod` enum specifies the encryption method:

| Value | Description |
|-------|-------------|
| `None` | No encryption. |
| `ZipCrypto` | Legacy ZIP encryption. Use only for backward compatibility—vulnerable to known-plaintext attacks. |
| `Aes128` | WinZip AES-128. |
| `Aes192` | WinZip AES-192. |
| `Aes256` | WinZip AES-256. **Recommended**—strongest available option. |

> | `Unknown` | Returned when the entry uses PKWare strong encryption, which .NET doesn't support. |

Reviewer: Strong encryption is only one of the examples that we don't support, but the description sounds more like it is the only one that leads to `Unknown` value. I would avoid mentioning specific algorithm names here.

### Choose AES-256 for new archives

When creating encrypted entries, always prefer `Aes256`. `ZipCrypto` is a legacy method with known cryptographic weaknesses and shouldn't be relied upon for security—use it only when interoperating with tools that don't support WinZip AES.

```csharp
// ⚠️ Weak encryption — use only for backward compatibility
archive.CreateEntry("file.txt", "password", ZipEncryptionMethod.ZipCrypto);

// ✅ Prefer AES-256
archive.CreateEntry("file.txt", "password", ZipEncryptionMethod.Aes256);
```

### Read encrypted entries

Use `ZipArchiveEntry.EncryptionMethod` to check the encryption method, and provide a password to `Open`:

```csharp
using ZipArchive archive = ZipFile.OpenRead("encrypted.zip");

foreach (ZipArchiveEntry entry in archive.Entries)
{
    if (entry.EncryptionMethod == ZipEncryptionMethod.Unknown)
    {

>         // PKWare strong encryption — not supported by .NET

Reviewer: Same as above, it may be some other unsupported algorithm

{
    Password = "myPassword".AsMemory(),
    EncryptionMethod = ZipEncryptionMethod.Aes256,
    CompressionLevel = CompressionLevel.Optimal
});
```

## Security checklist

Before deploying code that handles archives from untrusted sources, verify you've addressed each of the following:

- [ ] **Manual iteration:** Don't use `ExtractToDirectory` for untrusted input—iterate entries manually to enforce all limits.
- [ ] **Path traversal:** Validate all destination paths with `Path.GetFullPath()` + `StartsWith()`.
- [ ] **Decompression bombs:** Enforce limits on decompressed size (per-entry and total) and entry count.
- [ ] **Symlink/hardlink attacks (TAR):** Validate link targets resolve within the destination, or skip link entries entirely.
- [ ] **Memory limits:** Avoid `ZipArchiveMode.Update` for large untrusted archives. Avoid `Read` mode with unseekable streams from untrusted sources.
- [ ] **Thread safety:** Don't share `ZipArchive`, `TarReader`, or `TarWriter` instances across threads.
- [ ] **Untrusted metadata:** Treat entry names, comments, and extra fields as untrusted input. Sanitize before display or processing.

> - [ ] **File name validation:** On Windows, guard against reserved names (`CON`, `PRN`, `AUX`, `NUL`).

Reviewer: This is not mentioned elsewhere in the article?
